r1cs/lib/r1cs/Ops.cpp

```diff
@@ -19,62 +19,82 @@ using namespace mlir;
 namespace r1cs {
 
 ParseResult CircuitDefOp::parse(OpAsmParser &parser, OperationState &result) {
+  // Parse symbol name
   StringAttr symName;
   if (parser.parseSymbolName(symName, "sym_name", result.attributes)) {
     return failure();
   }
 
-  // Optional inputs (...)
   SmallVector<OpAsmParser::Argument> args;
-  Type argType;
+  DenseMap<unsigned, Attribute> perArgAttrs;
+
   if (succeeded(parser.parseOptionalKeyword("inputs"))) {
     if (parser.parseLParen()) {
       return failure();
     }
 
+    unsigned idx = 0;
     do {
       OpAsmParser::Argument arg;
-      if (parser.parseArgument(arg) || parser.parseColonType(argType)) {
+      Type type;
+      if (parser.parseArgument(arg) || parser.parseColonType(type)) {
         return failure();
       }
-      arg.type = argType;
+      arg.type = type;
+
+      // Try to parse an optional `{...}` attr
+      NamedAttrList attrList;
+      if (succeeded(parser.parseOptionalAttrDict(attrList))) {
+        if (auto pubAttr = attrList.get("pub")) {
+          perArgAttrs[idx] = pubAttr;
+        }
+      }
+
       args.push_back(arg);
+      idx++;
     } while (succeeded(parser.parseOptionalComma()));
 
     if (parser.parseRParen()) {
       return failure();
     }
   }
 
-  // Parse optional attribute dict
-  if (parser.parseOptionalAttrDict(result.attributes)) {
-    return failure();
+  // Later: attach DictionaryAttr mapping "0", "1", ... to parsed attr
+  NamedAttrList attrs;
+  for (auto [i, attr] : perArgAttrs) {
+    attrs.append(std::to_string(i), attr);
+  }
+
+  if (!attrs.empty()) {
+    result.addAttribute("arg_attrs", DictionaryAttr::get(parser.getContext(), attrs));
   }
 
-  // Parse body region with those args
   Region *body = result.addRegion();
   return parser.parseRegion(*body, args);
 }
 
-void CircuitDefOp::print(mlir::OpAsmPrinter &p) {
+void CircuitDefOp::print(OpAsmPrinter &p) {
   p << ' ';
   p.printSymbolName(getSymName());
 
   Block &entry = getBody().front();
+  bool hasAttrs = getArgAttrs().has_value();
+
   if (!entry.empty()) {
     p << " inputs (";
     llvm::interleaveComma(entry.getArguments(), p, [&](BlockArgument arg) {
       p << arg << ": ";
-      if (auto sigTy = arg.getType().dyn_cast<SignalType>()) {
-        sigTy.print(p);
-      } else {
-        p.printType(arg.getType()); // fallback for robustness
+      p.printType(arg.getType());
+
+      if (hasAttrs) {
+        auto dictAttr = getArgAttrs().value_or(DictionaryAttr::get(getContext()));
+        if (auto attr = dictAttr.get(std::to_string(arg.getArgNumber()))) {
+          p.printOptionalAttrDict(dictAttr.getValue());
+        }
```

Comment:
This doesn't seem right. Only `attr` be printed and not the entire dictionary.
Also, you could optimize by moving `dictAttr` out of the lambda.